### PR TITLE
Chunk stream

### PR DESCRIPTION
**Pull request by ilevyor**
*Description of changes:*
Added chunk-streak, checksum-stream, and checksums in order to support flexible checksums feature

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: include/aws/s3/private/s3_checksums.h [hunk 1]
```diff
@@ -0,0 +1,102 @@
+#ifndef AWS_S3_Checksums_H
```
Comment (bretambrose) [inline include/aws/s3/private/s3_checksums.h:1]: Caps

File: include/aws/s3/private/s3_checksums.h [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+#define AWS_S3_Checksums_H
+/**
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0.
+ */
+#include "aws/s3/s3_client.h"
+#include <aws/cal/hash.h>
+#include <aws/io/stream.h>
+
+#define AWS_CRC32_LEN 4
```
Comment (bretambrose) [inline include/aws/s3/private/s3_checksums.h:11]: Shouldn't these be in checksums itself?

File: include/aws/s3/private/s3_checksums.h [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+    struct aws_byte_buf *checksum_result);
+
+AWS_S3_API
+struct aws_input_stream *aws_chunk_stream_new(
+    struct aws_allocator *allocator,
+    struct aws_input_stream *existing_stream,
+    enum aws_s3_checksum_algorithm algorithm);
+
+AWS_S3_API
+size_t digest_size_from_algorithm(enum aws_s3_checksum_algorithm algorithm);
```
Comment (bretambrose) [inline include/aws/s3/private/s3_checksums.h:44]: "aws_get_" maybe

File: include/aws/s3/private/s3_checksums.h [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+
+/**
+ * Allocates and initializes a sha256 checksum instance.
+ */
+AWS_S3_API struct aws_checksum *aws_sha256_checksum_new(struct aws_allocator *allocator);
```
Comment (bretambrose) [inline include/aws/s3/private/s3_checksums.h:49]: These seem a little redundant given aws_checksum_new.  Also, we usually use a pattern ala "aws_checksum_new_sha256()" etc...

File: include/aws/s3/private/s3_checksums.h [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+ * Allocates and initializes a sha1 checksum instance.
+ */
+AWS_S3_API
+struct aws_checksum *aws_sha1_checksum_new(struct aws_allocator *allocator);
+/**
+ * Allocates and initializes an md5 checksum instance.
+ */
+AWS_S3_API
+struct aws_checksum *aws_md5_checksum_new(struct aws_allocator *allocator);
+/**
```
Comment (bretambrose) [inline include/aws/s3/private/s3_checksums.h:60]: Minor preference but I really like it when we put a newline between declaration-comment blocks.

File: include/aws/s3/private/s3_checksums.h [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+    void (*destroy)(struct aws_checksum *checksum);
+    int (*update)(struct aws_checksum *checksum, const struct aws_byte_cursor *buf);
+    int (*finalize)(struct aws_checksum *checksum, struct aws_byte_buf *out, size_t truncate_to);
+};
+
+struct aws_checksum {
+    struct aws_allocator *allocator;
+    struct aws_checksum_vtable *vtable;
+    void *impl;
+    size_t digest_size;
```
Comment (bretambrose) [inline include/aws/s3/private/s3_checksums.h:26]: Feels like it should be a vtable function instead.

File: include/aws/s3/private/s3_checksums.h [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+    bool good;
```
Comment (bretambrose) [inline include/aws/s3/private/s3_checksums.h:27]: Why is this necessary?

After looking further, I'd rename this based on its intent (calculation complete).

File: include/aws/s3/s3_client.h [hunk 8]
```diff
@@ -109,6 +109,25 @@ enum aws_s3_meta_request_compute_content_md5 {
     AWS_MR_CONTENT_MD5_ENABLED,
 };
 
+enum aws_s3_checksum_algorithm {
+    AWS_CRC32C,
```
Comment (bretambrose) [inline include/aws/s3/s3_client.h:113]: I would put a prefix on these: AWS_SOMETHING_ALGORITHM.  I usually just take the first letters of the word (SCA here) but anything reasonable should be fine.

File: source/s3_checksums.c [hunk 9]
```diff
@@ -0,0 +1,224 @@
...
+
+int s3_hash_finalize(struct aws_checksum *checksum, struct aws_byte_buf *output, size_t truncate_to) {
+    struct aws_hash *hash = (struct aws_hash *)checksum->impl;
+    checksum->good = false;
+    return aws_hash_finalize(hash, output, truncate_to);
+}
+
+struct aws_checksum *aws_hash_new(struct aws_allocator *allocator, aws_hash_new_fn hash_fn) {
+    struct aws_checksum *checksum = aws_mem_acquire(allocator, sizeof(struct aws_checksum));
+    struct aws_checksum_vtable *vtable = aws_mem_acquire(allocator, sizeof(struct aws_checksum_vtable));
```
Comment (bretambrose) [inline source/s3_checksums.c:42]: It's not a huge deal, but allocating a vtable on every object creation is not something we tend to do and isn't really necessary given that everything is single inheritance.

File: source/s3_checksums.c [hunk 10]
```diff
@@ -0,0 +1,224 @@
...
+
+void aws_crc_destroy(struct aws_checksum *checksum) {
+    aws_mem_release(checksum->allocator, checksum->vtable);
+    aws_mem_release(checksum->allocator, checksum);
+}
+
+uint32_t aws_crc32_common(
+    uint32_t previous,
+    const struct aws_byte_cursor *buf,
+    uint32_t (*checksum_fn)(const uint8_t *, int, uint32_t)) {
```
Comment (bretambrose) [inline source/s3_checksums.c:92]: feels like this could be typedefed

File: source/s3_checksums.c [hunk 11]
```diff
@@ -0,0 +1,224 @@
...
+        length -= (size_t)INT_MAX;
+    }
+    return checksum_fn(buffer, (int)length, val);
+}
+
+int aws_crc32_checksum_update(struct aws_checksum *checksum, const struct aws_byte_cursor *buf) {
+    if (!checksum->good) {
+        return aws_raise_error(AWS_ERROR_INVALID_STATE);
+    }
+    checksum->impl = (void *)(uintptr_t)aws_crc32_common((uint32_t)(uintptr_t)checksum->impl, buf, aws_checksums_crc32);
```
Comment (bretambrose) [inline source/s3_checksums.c:109]: This is deeply confusing. It appears that the impl for 32 bit checksums is the running checksum.  Looking at the tortuous casts, I would just make a simple impl that contains the actual value.

File: source/s3_chunk_stream.c [hunk 12]
```diff
@@ -0,0 +1,221 @@
...
+    if (aws_input_stream_get_length(existing_stream, &stream_length)) {
+        goto error3;
+    }
+    struct aws_byte_cursor pre_chunk_cursor = aws_byte_cursor_from_string(s_pre_chunk);
+    // Should a claculate the length here with a log algirthm?
+    // 2^64 is 20 digits long so if this overflows so do our int_64 lengths.
+    char stream_length_string[32];
+    AWS_ZERO_ARRAY(stream_length_string);
+    // #if _MSC_VER
+    // #    pragma warning(disable : 4996) /* sprintf */
```
Comment (bretambrose) [inline source/s3_chunk_stream.c:168]: Can remove

File: source/s3_chunk_stream.c [hunk 13]
```diff
@@ -0,0 +1,221 @@
...
+    // #endif
+    snprintf(stream_length_string, 31, "%" PRId64, stream_length);
```
Comment (bretambrose) [inline source/s3_chunk_stream.c:170]: AWS_ARRAY_SIZE(stream_length_string)

File: source/s3_chunk_stream.c [hunk 14]
```diff
@@ -0,0 +1,221 @@
...
+    int64_t post_trailer_len = s_post_trailer->len;
+    if (aws_input_stream_get_length(impl->current_stream, &prechunk_stream_len)) {
+        goto error;
+    }
+    impl->length = stream_length + prechunk_stream_len + final_chunk_len + post_trailer_len + encoded_checksum_len;
+    AWS_FATAL_ASSERT(impl->current_stream);
+    AWS_FATAL_ASSERT(impl->checksum_stream);
+    return stream;
+
+error:
```
Comment (bretambrose) [inline source/s3_chunk_stream.c:212]: You can remove all the extra error labels.  stream destroy checks for null.

Comment (ilevyor) [reply source/s3_chunk_stream.c:212]: Does it? I was getting funky memory issues due to accidentally destroying a stream more than once.
